[fuse] 가계부 #3 첫번째 PR

silvertae opened the pull request:
1일차: localStorage에 저장할 내역 데이터의 구조 설계 및 구현하기

localStorage에 데이터를 어떤 형식으로 저장할 것인지 고민만 하는데 시간을 너무 쓴 것 같습니다.
일단 렌더링부터 구현을 해봐야 후에 수정/삭제 하는 기능을 어떻게 만들지 감이 잡힐텐데 삽질만 했네요.
처음부터 너무 완벽하게 설계하려 하지 말고 일단 짜면서 고치기로 마음 먹었습니다.

2일차: 입력받은 데이터와 localStorage 활용해서 메인화면에 내역 렌더링하기

일단 submit 이벤트가 발생하는 순간 입력받은 데이터를 localStorage에 객체 형식으로 저장해두고
메인화면에 새로운 리스트를 렌더링하도록 구현해보고 있습니다.

3일차: 2일차에 이어서 렌더링 기능 구현중...

전날 남세 마스터의 코드리뷰를 받고 nullish 변환 연산자와 some, every 메소드에 대한 새롭게 학습도 하고 기존 코드도 수정해보았습니다.
 
그리고 어제에 이어 새롭게 입력받은 데이터를 기반으로 메인화면에 리스트를 렌더링하는 기능들을 구현 중입니다.
연월과 날짜별로 데이터들을 구분하여 상황에 맞게 내역들을 렌더링하는 것의 어려움을 느꼈습니다.
MVC 패턴등에 대해서도 찾아보았지만 아직은 직접 적용하기에는 너무 어려운 개념인 것 같습니다.

silvertae committed:
Fix: 확인 버튼 활성화 관련 파일 나누기 및 분류하기

submitBtnHandler와 관련된 함수들이 들어있던 파일을 작은 함수들로 나누고
분류를 다시 했다.

silvertae committed:
Feat: 파일들 변수명에서 $ 제거 및 date validation 재활성화

일관된 변수명 규칙을 사용하기 위해서 $ 기호를 앞에 쓴 변수들에서 기호를
제거하였다. 그리고 date validation 기능을 이용해 잘못된 형식의 날짜가
입력됐을 때 alert가 뜨게 하였다.

silvertae committed:
Docs: 3주차 학습 및 구현 계획 리드미에 업데이트

3주차 학습 및 구현 계획을 리드미 파일에 추가하였다. 아직 넣어야 할 기능
들이 많지만 우선 메인 화면 영역이라도 제대로 작동하도록 완성시키는 것이
현재 목표이다.

silvertae committed:
Chore: 내역을 입력받을 때 수입을 구분할 수 있는 income input 추가

인풋바에서 가격입력창에서 체크가 되어서 + 부호가 있을 때는 income 프로퍼티가
on으로 객체 형식으로 저장되는 데이터에 추가되도록 했다.

그리고 components 폴더에 들어있던 js 파일들을 밖으로 빼내고 폴더를 삭제했다.
마스터의 미션소개를 듣고 js 파일들을 나누는 기준을 정해서 다시 나눠보려 한다.

silvertae committed:
Feat: 내역 데이터를 로컬스토리지에 연월로 분리해서 저장하는 기능 추가

인풋바에서 입력받은 데이터를 후에 월별로 묶어서 관리하기 쉽게 하기위해,
같은 년도와 월을 가지는 데이터끼리 묶어서 로컬스토리지에 키-값 쌍으로
저장하도록 코드를 추가하였다.

silvertae committed:
Feat: 메인 화면에 새로운 내역리스트 만들어 렌더링하는 함수 추가

mainList.js 파일에 새로운 일자 내역리스트를 추가해서 렌더링하는 함수인
addNewList() 함수를 새로 만들었다. 후에 로컬스토리지에 저장된 데이터를
이용해 메인 화면에 내역을 렌더링 할 수 있도록 짜 볼 예정이다.

그리고 html 문서에서 불필요한 meta 태그를 지웠다.

silvertae committed:
Feat: 추가된 새로운 내역 세부사항 업데이트 기능 추가

새롭게 만들어진 일별 리스트에 입력받은 데이터를 바탕으로 각 항목별
세부사항을 입력해주는 함수인 addDailyDatail() 함수를 추가했다. 일단
submit 이벤트가 인풋바에서 발생할 때 동시에 추가되도록 연결시켜놓긴 했는데
테스트가 필요할 것 같다.

silvertae committed:
Fix: 내역 입력시 새로운 리스트 세부항목들이 렌더링 되지 않는 문제 수정

insertAdjacentElement와 insertAdjacentHTML을 착각해서 템플릿 리터럴로
만든 div 내용들이 제대로 리스트에 추가되지 않아서 렌더링되지 않는 문제를
수정하였다. 그리고 배열과 forEach를 사용해 중복되는 코드를 줄였다.

입력받은 데이터가 수입인지 지출인지에 따라 금액 색상이 다르게 나타나도록
클래스를 다르게 부여하는 코드를 추가하였다.

silvertae committed:
Docs: 3주차 미션 진행상황과 삽질기록 내용 추가

리드미 파일에 현재까지 진행상황을 체크리스트로 표현하고 어제 삽질했던
부분과 느낀 점을 정리해서 올렸다.

silvertae committed:
Chore: mainList 파일 간단히 수정

silvertae committed:
Docs: 리드미에 현재까지 구현한 것 미리보기 추가

namse commented on src/js/data/storeInputData.js:
```diff
@@ -0,0 +1,40 @@
+function processFormDate(data) {
+	const yearMonthId = data.date.slice(0, 6);
+	if (!localStorage.getItem(yearMonthId)) {
+		localStorage.setItem(yearMonthId, JSON.stringify({}));
```
이거 여기에 하는게 맞는가!?

namse replied:
const a = b ?? c; 같은 식이면 깔~끔 해질듯.

namse commented on src/js/inputbar/isAllFilled.js:
```diff
@@ -0,0 +1,7 @@
+export function isAllFilled() {
+	const inputs = document.querySelectorAll('form input[type=text]');
+	for (const input of inputs) {
```
Array.prototype.some, Array.prototype.every 라는 녀석이 있따능

silvertae replied:
위 코드를 `inputs.every(input => input.value)` 이런식으로 모든 인풋값이 있는 경우만 true를 반환하려고 했는데, querySelectorAll로 반환되는 리턴값이 Array가 아니라 NodeList라 some 메소드를 사용할 수 없는 것 같습니다.
NodeList를 Array로 변환해서라도 some, every를 사용하는 것이 좋을까요?

namse replied:
그런 이유가 있었구만~ 아뇨 굳이 안그래도 돼요

silvertae replied:
감사합니다. 그래도 some, every 라는 method가 있다는 것을 새롭게 알게 되었네요.

silvertae committed:
Fix: PR 리뷰 후 피드백 받은 부분에 Nullish coalescing operator 적용

코드 리뷰때 피드백 받은 부분의 코멘트를 읽어보고 ?? 라는 널 병합 연산자가
존재한다는 것을 처음 알게 되었다. 일반적인 OR 연산자(||)와 비슷해 보이지만
null이나 undefined 일때만 우측 피연산자를 반환한다는 점이 달랐다.

지금 코드에 어떻게 적용할 수 있을까 고민해보다가, 어차피 뒤에 호출한
함수에서 localStorage에 monthlyList를 setItem 해주니
굳이 processFormDate() 함수에서 localStorage에 key값이 없을때
빈 객체를 미리 넣어둘 필요가 없다는 것을 알게 되었다.

silvertae committed:
Docs: 널 병합 연산자에 대한 학습정리 내용 추가

리드미 파일에 널 병합 연산자(Nullish coalescing operator)에 관해서
학습한 내용을 정리해서 추가하였다.

silvertae committed:
Merge branch 'feature/storage' into develop

silvertae committed:
Docs: MVC 패턴에 대해서 조사한 내용을 추가

MVC 패턴과 이를 FE에서는 어떤 방식으로 적용하는지에 대해서 간단히
조사해보고 리드미 학습정리에 추가하였다. 지금 당장 이해하고 적용하기에는
무리가 있다고 판단해서 우선 짧게 정리만 해두었다.

silvertae committed:
Feat: 메인화면 렌더링 시 기존 날짜면 새로운 데일리 리스트를 안 만드는 기능 추가

isNewDate() 함수로 메인화면을 렌더링 할 때 기존 로컬 스토리지에
해당하는 날짜의 데이터가 이미 존재하는 경우에는 세부내역 리스트만 추가하고,
새로운 날짜인 경우는 daily-list를 새로 만들도록 구분하는 기능을 추가했다.

근데 현재는 페이지 처음 로드시 메인 화면이 초기화 돼서 처음 로드시
로컬스토리지 데이터를 바탕으로 렌더링하는 기능이 추가되어야 할 것 같다.

silvertae committed:
Merge branch 'feature/storage' into develop

silvertae committed:
Chore: 새로 추가되는 daily-list에 구별하기 위한 id 속성 추가

새로운 날짜의 내역이 입력되었을 때 새롭게 생성되는 daily-list 내부에
존재하는 daily-detail-lists-container라는 클래스를 가지는 ul 태그에
입력받은 데이터의 date값을 이용한 고유 id를 부여하고 이를 기반으로
새로운 세부내역들을 추가하도록 하였다.

후에 날짜별 리스트를 구분하려면 id가 필요하다고 생각했다.

silvertae committed:
Feat: 새로 추가된 날짜 내역에 daily-info 렌더링 하는 기능 추가

새롭게 추가된 daily-list의 경우 함께 그날의 날짜와 요약을 보여주는
daily-info란을 렌더링 하도록 해주는 함수 addDailyInfo()를 추가했다.

이를 업데이트 해줄 updateDailyInfo() 함수도 만들어 추가예정이다.

lolWK commented on README.md:
```diff
@@ -231,12 +249,71 @@
 
 - css의 :root 파일에서 font 한줄 선언 → [https://stackoverflow.com/questions/4218549/one-css-declaration-for-all-css-font-properties](https://stackoverflow.com/questions/4218549/one-css-declaration-for-all-css-font-properties)
 
+## 📚 3주차 학습 정리
```
학습 정리 멋져요 퓨즈 👍

lolWK commented on src/js/header/displayCurrentMonthYear.js:
```diff
@@ -0,0 +1,16 @@
+	const currentMonthEl = document.querySelector('div.current-month-year p:nth-of-type(2)');
+	const currentMonthNameEl = document.querySelector('div.current-month-year p:nth-of-type(3)');
+
+	currentYearEl.innerText = currentYear;
```
innerText보다 textContent를 사용하는 것이 좋지 않을까요? 🤔

silvertae replied:
앗 바꿔야지 했는데 깜빡했네요. 감사합니다!

lolWK commented on index.html:
```diff
@@ -247,55 +245,55 @@ <h1 class="logo-area">
 						<div class="daily-info">
-							<div><strong>3월 17일</strong> 수</div>
-							<div>지출 <span>27,000</span></div>
+							<div><span>3월 17일</span> <span>수</span></div>
```
strong태그에서 잘 바꾼 것 같아요!

silvertae replied:
처음에 html 구조 만들때 날짜만 색이 다르길래 구분하려고 strong 태그를 썼었는데 굳이 필요없었던 것 같아요
